main.cpp

```diff
@@ -294,256 +242,153 @@ struct IntType;
-    DoubleType& operator/=( double rhs );
-
-    DoubleType& pow(double rhs);
-    DoubleType& pow(const FloatType& rhs);
-    DoubleType& pow(const DoubleType& rhs);
-    DoubleType& pow(const IntType& rhs);
-
-    DoubleType& apply(std::function<DoubleType&(double&)> dtLamb);
-    DoubleType& apply(void(*dtPtr)(double&));
-
-    operator double() const { return *value; }
+    operator Type() const { return *value; }
 
 private:
-    double* value = nullptr;
-    DoubleType& powInternal(double arg);
+    std::unique_ptr<Type> value;
+    Numeric& powInternal(Type arg)
+    {
+        *value = static_cast<Type>(std::pow( *value, arg ));
+        return *this;
+    }
 };
 
-DoubleType& DoubleType::apply(std::function<DoubleType&(double&)> dtLamb)
+template<typename NumericType>
+void myNumericFreeFunct(std::unique_ptr<NumericType>& ftPtr)
 {
-    if ( dtLamb )
-        return dtLamb(*value);
-    return *this;
+    NumericType& ftFreeFunc = *ftPtr;
+    ftFreeFunc += static_cast<NumericType>(7.0);
 }
 
-DoubleType& DoubleType::apply(void(*dtPtr)(double&))
+template<>
+struct Numeric<double>
 {
-    if( dtPtr )
-        dtPtr(*value);
-    return *this;
-}
+    using Type = double;
 
-DoubleType& DoubleType::pow(double rhs)
-{
-    return powInternal(rhs);
-}
+    explicit Numeric(Type ft) : value( std::make_unique<Type>(ft) ) { }
 
-DoubleType& DoubleType::powInternal(double arg)
-{
-    *value = std::pow( *value, arg );
-    return *this;
-}
-
-DoubleType& DoubleType::operator+=( double rhs )
-{
-    *value += rhs;
-    return *this;
-}
-DoubleType& DoubleType::operator-=( double rhs )
-{
-    *value -= rhs;
-    return *this;
-}
-DoubleType& DoubleType::operator*=( double rhs )
-{
-    *value *= rhs;
-    return *this;
-}
-DoubleType& DoubleType::operator/=( double rhs )
-{
-    if (rhs == 0.0)
-        std::cout << "warning: floating point division by zero!" << "\n";
-        
-    *value /= rhs;
-    return *this;
-}
+    Numeric& operator+=(const Type& rhs)
+    {
+        *value += rhs;
+        return *this;
+    }
 
-void myDoubleFreeFunct(double& dtPtr)
-{
-    double& dtFreeFunc = dtPtr;
-    dtFreeFunc += 6.0;
-}
+    Numeric& operator-=(const Type& rhs)
+    {
+        *value -= rhs;
+        return *this;
+    }
 
-struct IntType
-{
-    explicit IntType(int it) : value(new int(it)) {}
-    ~IntType()
+    Numeric& operator*=(const Type& rhs)
     {
-        delete value;
-        value = nullptr;
+        *value *= rhs;
+        return *this;
     }
 
-    IntType& operator+=( int rhs );
-    IntType& operator-=( int rhs );
-    IntType& operator*=( int rhs );
-    IntType& operator/=( int rhs );
+    template<typename DividedByType>
+    Numeric& operator/=(const DividedByType& rhs)
+    {
+        if (std::abs(rhs) <= std::numeric_limits<DividedByType>::epsilon())
+            std::cout << "warning: floating point division by zero!" << "\n";
+        *value /= rhs;
+        return *this;
+    }
```

Comment:
- This does not check if `rhs == 0.0`
- It is also not using `const Type&` as the argument type.
- [ ] revise